### PR TITLE
RoundSystem: Added polymorphic round rework

### PR DESCRIPTION
**Pull request by mzegar**
Closed previous PR/fork due to random file generation repos... https://github.com/TTTReborn/ttt-reborn/pull/4

**Changes**
- Adds polymorphic round rework
- Rounds properly cycle between 
   - Waiting -> Pre -> InProgress -> Post

### CODE REVIEW

File: code/ttt-reborn.csproj [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (mzegar) [inline code/ttt-reborn.csproj:1]: This file is not from IDE, but from s&box project generation file.

Comment (Alf21) [reply code/ttt-reborn.csproj:1]: I don't think we need that file as well as we already have a `tttreborn.csproj` file. Not matching the repo's name, but the gamemode's name. If you'd change the add-on's folder into `tttreborn` as well, the generation tool would name this file like mine. But we can use your file as well, it doesn't matter, just don't need to have two of the same files :D

File: code/rounds/WaitingRound.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        protected override void OnStart()
+        {
+            if (Host.IsServer)
+            {
+                foreach (Client client in Client.All)
+                {
+                    if (client.Pawn is TTTPlayer player)
+                        player.Respawn();
```
Comment (Alf21) [inline code/rounds/WaitingRound.cs:46]:
```suggestion
                    if (client.Pawn is TTTPlayer player)
                    {
                        player.Respawn();
                    }
```

File: code/rounds/PreRound.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public override int RoundDuration => TTTReborn.Gamemode.Game.TTTPreRoundTime;
+
+        protected override void OnStart()
+        {
+            if (Host.IsServer)
+            {
+                foreach (Client client in Client.All)
+                {
+                    if (client.Pawn is TTTPlayer player)
+                        player.Respawn();
```
Comment (Alf21) [inline code/rounds/PreRound.cs:20]:
```suggestion
                    if (client.Pawn is TTTPlayer player)
                    {
                        player.Respawn();
                    }
```

File: code/rounds/PostRound.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        protected override void OnTimeUp()
+        {
+            TTTReborn.Gamemode.Game.Instance.ChangeRound(new PreRound());
+
+            base.OnTimeUp();
+        }
+
+        public override void OnPlayerSpawn(TTTPlayer player)
+        {
+            if (Players.Contains(player)) return;
```
Comment (Alf21) [inline code/rounds/PostRound.cs:19]:
```suggestion
            if (Players.Contains(player))
            {
                return;
            }
```

File: code/rounds/InProgressRound.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            bool traitorsDead = true;
+            bool innocentsDead = true;
+
+            // Check for alive players
+            for (int i = 0; i < Client.All.Count; i++)
+            {
+                TTTPlayer alivePlayer = Client.All[i].Pawn as TTTPlayer;
+
+                if (alivePlayer.LifeState == LifeState.Alive)
+                    continue;
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:94]:
```suggestion
                if (alivePlayer.LifeState == LifeState.Alive)
                {
                    continue;
                }
```

File: code/rounds/InProgressRound.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        private async Task ChangeToPostRound(int delay = 3)
+        {
+            _isGameOver = true;
+
+            await Task.Delay(delay * 1000);
+
+            if (TTTReborn.Gamemode.Game.Instance.Round != this)
+                return;
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:78]:
```suggestion
            if (TTTReborn.Gamemode.Game.Instance.Round != this)
            {
                return;
            }
```

File: code/rounds/InProgressRound.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            if (Host.IsServer)
+            {
+                Spectators.Clear();
+            }
+        }
+
+        protected override void OnTimeUp()
+        {
+            if (_isGameOver) return;
```
Comment (Alf21) [inline code/rounds/InProgressRound.cs:54]:
```suggestion
            if (_isGameOver)
            {
                return;
            }
```

File: code/Properties/launchSettings.json [hunk 8]
```diff
@@ -1,9 +1,9 @@
...
-        }
+  "profiles": {
+    "ttt-reborn": {
+      "commandName": "Executable",
+      "executablePath": "C:\\Program Files (x86)\\Steam\\steamapps\\common\\sbox\\sbox.exe",
+      "commandLineArgs": "\u002Bdeveloper \u002Bsw -noassert -tools"
     }
-}
+  }
+}
```
Comment (Alf21) [inline code/Properties/launchSettings.json:9]: This isn't needed as well :D

File: code/Game.cs [hunk 9]
```diff
@@ -232,18 +56,50 @@ public override void ClientJoined(Client client)
...
+    }
+
+    private async Task StartGameTimer()
+    {
+        ChangeRound(new WaitingRound());
+        while (true)
+        {
+            await Task.DelaySeconds(1);
+            OnGameSecond();
+        }
```
Comment (Alf21) [inline code/Game.cs:87]:
```suggestion
        ChangeRound(new WaitingRound());

        while (true)
        {
            await Task.DelaySeconds(1);

            OnGameSecond();
        }
```

File: code/Game.cs [hunk 10]
```diff
@@ -198,28 +34,16 @@ public override void DoPlayerSuicide(Client client)
...
-
     public override void OnKilled(Entity entity)
     {
-        Client client = entity.GetClientOwner();
-        if (client != null)
-        {
-            CheckRoundState();
-        }
+        if (entity is TTTPlayer player)
+            Round?.OnPlayerKilled(player);
```
Comment (Alf21) [inline code/Game.cs:40]:
```suggestion
        if (entity is TTTPlayer player)
        {
            Round?.OnPlayerKilled(player);
        }
```

**Review by Alf21 — CHANGES_REQUESTED** [review]
First of all, thanks again for these changes and the implementation, great addition and well matching our coding style like the coding convention. Anyways, some small changerequests. We decided not using "one line if-conditions" without brackets as this could lead to carelessness mistakes if you wanna add something to that condition for debugging etc. It does not cost a lot of time making it constent, but reducing issues :D

File: code/ttt-reborn.csproj (hunk 1 above)
Comment (mzegar) [reply code/ttt-reborn.csproj:1]: Ah, didn't realize it was based on folder name. Assumed it was based off of `.addon` or something similar.

File: code/rounds/PreRound.cs [hunk 11]
```diff
@@ -0,0 +1,58 @@
+using Sandbox;
+using System.Threading.Tasks;
+
+using TTTReborn.Player;
+
+namespace TTTReborn.Rounds
+{
+    public class PreRound: BaseRound
+    {
+        public override string RoundName => "Pre";
```
Comment (Alf21) [inline code/rounds/PreRound.cs:10]:
```suggestion
        public override string RoundName => "Preparing";
```

File: code/rounds/PreRound.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            TTTReborn.Gamemode.Game.Instance.ChangeRound(new InProgressRound());
+
+            base.OnTimeUp();
+        }
+
+        private async Task StartRespawnTimer(TTTPlayer player)
+        {
+            await Task.Delay(1000);
+
+            player.Respawn();
```
Comment (Alf21) [inline code/rounds/PreRound.cs:43]: Here, we need a check whether the currently active round is still the PreRound as the delay could result in a player respawn of another (InProgress) round

File: code/rounds/WaitingRound.cs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                }
+            }
+        }
+
+        private async Task StartRespawnTimer(TTTPlayer player)
+        {
+            await Task.Delay(1000);
+
+            player.Respawn();
```
Comment (Alf21) [inline code/rounds/WaitingRound.cs:55]: Theoretically, same as above, but not critical here

**Review by Alf21 — CHANGES_REQUESTED** [review]

File: code/ttt-reborn.csproj (hunk 1 above)
Comment (Alf21) [reply code/ttt-reborn.csproj:1]: Alright, then we have to remove `tttreborn.csproj` here as well :D

File: code/Properties/launchSettings.json (hunk 8 above)
Comment (mzegar) [reply code/Properties/launchSettings.json:9]: I don't think the `.gitignore` will ignore until the file is committed. So will need to delete this file after the merge.

File: code/rounds/WaitingRound.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                    }
+                }
+            }
+        }
+
+        private async Task StartRespawnTimer(TTTPlayer player)
+        {
+            await Task.Delay(1000);
+
+            if (TTTReborn.Gamemode.Game.Instance.Round is WaitingRound)
```
Comment (mzegar) [inline code/rounds/WaitingRound.cs:57]: Might not even be needed, or could even use `this` instead. Would test but my IDE keeps fighting will me, hopefully https://github.com/TTTReborn/ttt-reborn/pull/8 will help standardize a lot of things.

**Review by Alf21 — APPROVED** [review]
😊👍

File: code/ttt-reborn.csproj [hunk 15]
```diff
@@ -3,8 +3,8 @@
 	<PropertyGroup>
 		<TargetFramework>net5.0</TargetFramework>
 		<GenerateDocumentationFile>true</GenerateDocumentationFile>
-		<AssemblyName>tttreborn</AssemblyName>
-		<PackageId>tttreborn</PackageId>
+		<AssemblyName>ttt-reborn</AssemblyName>
+		<PackageId>ttt-reborn</PackageId>
```
Comment (TimGoll) [inline code/ttt-reborn.csproj:7]: why is this changed back? Didn't Alf just change it from `ttt-reborn` to `tttreborn`

Comment (mzegar) [reply code/ttt-reborn.csproj:7]: We decided on marking any .csproj files on the gitignore since they are unique to each machine.

From my understanding of gitignore it'll start ignoring files once it's commited to repo. So will delete file once merged.

Comment (TimGoll) [reply code/ttt-reborn.csproj:7]: oh, alright. I missed it, sorry!

**Review by TimGoll — APPROVED** [review]
